### PR TITLE
Feature Quantization [3/5]: Preprocessor (write-side) Updates

### PR DESCRIPTION
**Pull request by jchmura-sc**
**Purpose**
The purpose of this PR is to make the (write-side) preprocessor changes for feature quantization:
- `FeatureQuantizationSpec` the API endpoint that the user can specific inside `NodeDataPreprocessingSpec`
- beam transform for computing quantization metadata and serializing to temporary json
- beam transform for quantizing node features based on that metadata (calls into `quantize_ndarray`)
- driver reads the json, constructs `FeatureQuantizationMetadata` proto, and adds to `preprocessed_metadata_pb`

**Context**
- #733
- #734
- #735
- #736 
- #737 

***Updated Changelog.md?*** NO

***Ready for code review?:*** NO

### CODE REVIEW

File: gigl/src/data_preprocessor/lib/transform/feature_quantization.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    packed_feature.value_count.max = 1
+    logger.info(
+        f"Updated transformed schema for feature quantization: dropped "
+        f"{len(spec.feature_keys)} features and added bytes feature "
+        f"{_NODE_PACKED_FEATURE_KEY}."
+    )
+    return quantized_schema
+
+
+def _build_feature_values(
```
Comment (jchmura-sc) [inline gigl/src/data_preprocessor/lib/transform/feature_quantization.py]: TBD: Confirming we impact of switching to asymetric clip values. If we proceed then this can be removed

File: gigl/src/data_preprocessor/lib/transform/feature_quantization.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+from gigl.common.utils.feature_quantization.numpy_ops import quantize_ndarray
+from gigl.common.utils.tensorflow_schema import feature_spec_to_feature_index_map
+from gigl.src.data_preprocessor.lib.types import FeatureQuantizationSpec
+
+logger = Logger()
+_NODE_PACKED_FEATURE_KEY: Final[str] = "node_packed_features"
+_SignStats: TypeAlias = tuple[float, int, float, int]
+
+
+def apply_feature_quantization_transform(
```
Comment (kmontemayor2-sc) [inline gigl/src/data_preprocessor/lib/transform/feature_quantization.py:22]: can we add docstrings here?

File: gigl/src/data_preprocessor/lib/transform/feature_quantization.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    return quantized_schema
+
+
+def _flatten_feature_values(
+    batch: pa.RecordBatch, quantized_feature_keys: list[str]
+) -> list[float]:
+    return _build_feature_matrix(batch, quantized_feature_keys).ravel().tolist()
+
+
+def _build_feature_matrix(
```
Comment (kmontemayor2-sc) [inline gigl/src/data_preprocessor/lib/transform/feature_quantization.py:235]: BTW do we have concerns going back and forth from pa to np?

File: gigl/src/data_preprocessor/lib/transform/feature_quantization.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    batch: pa.RecordBatch, quantized_feature_keys: list[str]
+) -> np.ndarray:
+    key_to_idx: dict[str, int] = {name: i for i, name in enumerate(batch.schema.names)}
```
Comment (kmontemayor2-sc) [inline gigl/src/data_preprocessor/lib/transform/feature_quantization.py:238]: Does this require is to have the same batch creation across all machines? does pyarrow / beam take care of this?

File: gigl/src/data_preprocessor/lib/transform/feature_quantization.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+
+def _multi_bit_stats_from_quantiles(quantiles: list[float]) -> dict[str, float]:
+    if not quantiles:
+        raise ValueError("Cannot compute quantization stats from no values.")
+    quantile_count = len(quantiles) - 1
+    clip_min = float(quantiles[round(0.005 * quantile_count)])
+    clip_max = float(quantiles[round(0.995 * quantile_count)])
+    if clip_max <= clip_min:
+        clip_max = clip_min + 1e-5
```
Comment (kmontemayor2-sc) [inline gigl/src/data_preprocessor/lib/transform/feature_quantization.py:265]: Is this case for when the data are all clustered tight enough together that you can't really distinguish the quantiles?

File: gigl/src/data_preprocessor/lib/transform/utils.py [hunk 6]
```diff
@@ -362,6 +366,26 @@ def get_load_data_and_transform_pipeline_component(
...
+                apply_feature_quantization_transform(
+                    logical_features=transformed_features,
+                    logical_metadata=logical_metadata,
+                    logical_feature_keys=list(
+                        preprocessing_spec.features_outputs or []
+                    ),
+                    quantization_spec=quantization_spec,
+                    quantization_metadata_path=transformed_features_info.feature_quantization_metadata_path.uri,
+                )
+            )
```
Comment (kmontemayor2-sc) [inline gigl/src/data_preprocessor/lib/transform/utils.py:388]: Does this always launch the quantization pipeline? Even if we don't want to quantize anything?

File: gigl/src/data_preprocessor/lib/types.py [hunk 7]
```diff
@@ -48,6 +49,20 @@ class NodeOutputIdentifier(str):
...
+class FeatureQuantizationSpec:
+    """Selects logical feature fields to pack at a fixed bit width."""
+
+    feature_keys: list[str]
+    bits: int
+
+    def __post_init__(self) -> None:
+        if not self.feature_keys:
+            raise ValueError("Feature quantization expects at least one feature key.")
+        if self.bits not in (1, 2, 4, 8):
```
Comment (kmontemayor2-sc) [inline gigl/src/data_preprocessor/lib/types.py]: nit. let's consider adding some SUPPORTED_QUANTIZATION_BITS in the utils folder so we can reference anywhere (and easier to update if needed)

**Review by kmontemayor2-sc — COMMENTED** [review]
Thanks Jacob!

File: gigl/src/data_preprocessor/lib/transform/feature_quantization.py (hunk 5 above)
Comment (jchmura-sc) [reply gigl/src/data_preprocessor/lib/transform/feature_quantization.py:265]: Yeah exactly - I can imagine this happening in case a feature happens to be fully zero-d out for instance. This prevents issues in the [min-max scaling](https://github.com/Snapchat/GiGL/blob/b9558cf3d4a10e22107ab36e832f475e68a6fcf4/gigl/common/utils/feature_quantization/numpy_ops.py#L40)

File: gigl/src/data_preprocessor/lib/transform/utils.py (hunk 6 above)
Comment (jchmura-sc) [reply gigl/src/data_preprocessor/lib/transform/utils.py:388]: If we don't want to quantize anything then `FeatureQuantizationSpec` will be `None` (which is automatically true for non-node preprocessing spec), and the quantization pipeline is skipped.

Comment (jchmura-sc) [reply gigl/src/data_preprocessor/lib/transform/utils.py:388]: And if we happen to instantiate the spec with an empty list of feature to quantize, we fail loudly during construction of the `FeatureQuantizationSpec`

File: gigl/src/data_preprocessor/lib/types.py (hunk 7 above)
Comment (jchmura-sc) [reply gigl/src/data_preprocessor/lib/types.py]: Good idea, will do

File: gigl/src/data_preprocessor/lib/transform/feature_quantization.py (hunk 3 above)
Comment (jchmura-sc) [reply gigl/src/data_preprocessor/lib/transform/feature_quantization.py:235]: Performance wise? I intentionally allow `zero_copy_only=False` which  can fallback to creating a copy in the presence of null values (though this triggers an exception later anyway). In case data is good this should be zero copy.

More broadly, I suspect that row-wise ops (i.e. bit-packing) which we do during quantization would not be possible to do in pyarrow compute

File: gigl/src/data_preprocessor/lib/transform/feature_quantization.py (hunk 4 above)
Comment (jchmura-sc) [reply gigl/src/data_preprocessor/lib/transform/feature_quantization.py:238]: No, I don't think so.

The feature matrix is ultimately constructed in the `FeatureQuantizationSpec.feature_keys` order, regardless of per-worker batch schema order - they should all ultimately produces the exact same logical feature order.

But we do require that every batch contains every configured feature name in the quantizatino spec.

File: proto/snapchat/research/gbml/preprocessed_metadata.proto
Comment (kmontemayor2-sc) [inline proto/snapchat/research/gbml/preprocessed_metadata.proto:33]: should we re-base this on main or something? I thought we added these already?

**Review by kmontemayor2-sc — COMMENTED** [review]
Nice work, Jacob!

File: gigl/common/utils/feature_quantization/__init__.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+"""Utilities for node feature quantization in GiGL."""
+
+from typing import Final
+
+SUPPORTED_QUANTIZATION_BITS: Final[tuple[int, ...]] = (1, 2, 4, 8)
```
Comment (mkolodner-sc) [inline gigl/common/utils/feature_quantization/__init__.py]: nit: I'd have a preference for moving constants-responsible to a different file rather than declaring them inside of `__init__`. if we'd want to import from `gigl.common.utils.feature_quantization` we can then add import from the constants file here

File: gigl/src/data_preprocessor/lib/transform/feature_quantization.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+    Stores the packed bytes in ``node_packed_features`` and computes global
+    quantization statistics with Beam.
+
+    Side Effects:
+        Writes the quantization statistics JSON that ``data_preprocessor.py``
+        reads and serializes into the preprocessing metadata protobuf.
+
+    Args:
+        logical_features: RecordBatches containing the logical feature columns.
+        logical_metadata: Eager or deferred metadata for the logical schema.
```
Comment (mkolodner-sc) [inline gigl/src/data_preprocessor/lib/transform/feature_quantization.py]: nit: can you elaborate what eager or deferred mean here

File: gigl/src/data_preprocessor/lib/types.py [hunk 10]
```diff
@@ -48,6 +50,22 @@ class NodeOutputIdentifier(str):
...
 
+@dataclass(frozen=True)
+class FeatureQuantizationSpec:
+    """Selects logical feature fields to pack at a fixed bit width."""
+
+    feature_keys: list[str]
+    bits: int
+
+    def __post_init__(self) -> None:
+        if not self.feature_keys:
```
Comment (mkolodner-sc) [inline gigl/src/data_preprocessor/lib/types.py:61]: Is there any edge case behavior if there are duplicate feature keys in this list?

File: tests/integration/pipeline/data_preprocessor/feature_quantization_transform_test.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+            ),
+            (
+                "single_bit",
+                1,
+                [(-4.0, -2.0), (4.0, 8.0)],
+                {"neg_mean": -3.0, "pos_mean": 6.0},
+            ),
+        ]
+    )
+    def test_apply_feature_quantization_transform_writes_metadata(
```
Comment (mkolodner-sc) [inline tests/integration/pipeline/data_preprocessor/feature_quantization_transform_test.py:40]: Should we add a test for the eager/deferred branches here as well?

**Review by mkolodner-sc — APPROVED** [review]
Thanks Jacob!

File: gigl/src/data_preprocessor/lib/types.py (hunk 10 above)
Comment (jchmura-sc) [reply gigl/src/data_preprocessor/lib/types.py:61]: We have a guard on the read side but I'll add add a check here to surface the error fast

File: tests/integration/pipeline/data_preprocessor/feature_quantization_transform_test.py (hunk 11 above)
Comment (jchmura-sc) [reply tests/integration/pipeline/data_preprocessor/feature_quantization_transform_test.py:40]: Good point, will add

**Review by yliu2-sc — APPROVED** [review]
Stamp

File: gigl/src/data_preprocessor/data_preprocessor.py [hunk 12]
```diff
@@ -481,6 +482,31 @@ def generate_preprocessed_metadata_pb(
                 feature_dim=feature_dim_output,
                 transform_fn_assets_uri=node_transformed_features_info.transformed_features_transform_fn_assets_path.uri,
             )
+            metadata_path = (
+                node_transformed_features_info.feature_quantization_metadata_path.uri
+            )
+            if tf.io.gfile.exists(metadata_path):
```
Comment (kmontemayor2-sc) [inline gigl/src/data_preprocessor/data_preprocessor.py]: Hmmm, if the uri is set but doesn't exist, should that be an error?

File: tests/integration/pipeline/data_preprocessor/feature_quantization_transform_test.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+                        | "Create RecordBatches" >> beam.Create(batches),
+                        logical_metadata=logical_metadata,
+                        logical_feature_keys=logical_feature_keys,
+                        quantization_spec=FeatureQuantizationSpec(
+                            feature_keys=logical_feature_keys, bits=bits
+                        ),
+                        quantization_metadata_path=metadata_path,
+                    )
+                )
+                if use_deferred_metadata:
```
Comment (kmontemayor2-sc) [inline tests/integration/pipeline/data_preprocessor/feature_quantization_transform_test.py:92]: nit. I prefer to not have branching logic like this in tests, but I guess it's hard to do otherwise in this case?

**Review by kmontemayor2-sc — APPROVED** [review]

File: gigl/src/data_preprocessor/data_preprocessor.py (hunk 12 above)
Comment (jchmura-sc) [reply gigl/src/data_preprocessor/data_preprocessor.py]: Agreed, this shouldn't happen, but currently would silently skip quantization. I'll update

File: tests/integration/pipeline/data_preprocessor/feature_quantization_transform_test.py (hunk 13 above)
Comment (jchmura-sc) [reply tests/integration/pipeline/data_preprocessor/feature_quantization_transform_test.py:92]: I don't think its straightforward to do without additional indirection or a duplicate test
